File: scripts/changelog.ts
/** Работа с CHANGELOG.md: раздел «Не выпущено» и заметки к выпуску. */

export const UNRELEASED = "## [Не выпущено]";

/** Текст раздела версии (без заголовка) или undefined. */
export function section(changelog: string, version: string): string | undefined {
  const head = version === "unreleased" ? UNRELEASED : `## [${version}]`;
  const start = changelog.indexOf(head);
  if (start < 0) return undefined;
  const bodyStart = changelog.indexOf("\n", start) + 1;
  const next = changelog.slice(bodyStart).search(/^## \[/m);
  const body = next < 0 ? changelog.slice(bodyStart) : changelog.slice(bodyStart, bodyStart + next);
  return body.replace(/^\[[^\]]+\]:.*$/gm, "").trim();
}

/** Превращает «Не выпущено» в раздел версии и открывает новый пустой. */
export function cut(changelog: string, version: string, date: string): string {
  const body = section(changelog, "unreleased");
  if (!body) throw new Error("раздел «Не выпущено» пуст или отсутствует: нечего выпускать");
  return changelog.replace(UNRELEASED, `${UNRELEASED}\n\n## [${version}] — ${date}`);
}

File: scripts/changelog.py
"""Работа с CHANGELOG.md: раздел «Не выпущено» и заметки к выпуску."""

from __future__ import annotations

import re

UNRELEASED = "## [Не выпущено]"

NEXT_SECTION = re.compile(r"^## \[", re.MULTILINE)
LINK_DEFINITION = re.compile(r"^\[[^\]]+\]:.*$", re.MULTILINE)


def section(changelog: str, version: str) -> str | None:
    """Текст раздела версии (без заголовка) или None."""
    head = UNRELEASED if version == "unreleased" else f"## [{version}]"
    start = changelog.find(head)
    if start < 0:
        return None
    body_start = changelog.find("\n", start) + 1
    rest = changelog[body_start:]
    m = NEXT_SECTION.search(rest)
    body = rest if m is None else rest[: m.start()]
    return LINK_DEFINITION.sub("", body).strip()


def cut(changelog: str, version: str, date: str) -> str:
    """Превращает «Не выпущено» в раздел версии и открывает новый пустой."""
    if not section(changelog, "unreleased"):
        raise ValueError("раздел «Не выпущено» пуст или отсутствует: нечего выпускать")
    return changelog.replace(UNRELEASED, f"{UNRELEASED}\n\n## [{version}] — {date}", 1)
